bpf: Track reply packets of egress gateway and SNATed host-local connections

gentoo-root opened the pull request:
Flag the egress gateway NAT entries as needs_ct, so that reply packets could extend the expiration timeout as well.

Fix the conntrack tuple matching for packets in the reply direction, so that the needs_ct flag could actually be useful. Currently, the tuple before revSNAT is used for conntrack and doesn't match the original CT entry. After the fix, the tuple is adjusted to contain data like after revSNAT, so that it could match the original CT entry and get the needs_ct flag from it.

Fixes: #25110

```release-note
Track reply packets in long-living egress gateway connections and SNATed host-local connections.
```

gentoo-root committed:
bpf: Rename host_local to needs_ct in struct nat_entry

This flag will be reused for more cases that need connection tracking.
Currently it's only used for host-local connections, see commit
0e785f96756b ("bpf: initial bpf-based masquerading support") for
details.

Signed-off-by: Maxim Mikityanskiy <maxim@isovalent.com>

gentoo-root committed:
bpf: Track egress gateway reply packets

Flag the egress gateway NAT entries as needs_ct, so that reply packets
could extend the expiration timeout as well.

The next commit is also required for the complete fix, otherwise the
needs_ct mark won't be matched.

Signed-off-by: Maxim Mikityanskiy <maxim@isovalent.com>

gentoo-root committed:
bpf: Use the right tuple to track connections on revSNAT

Fix a bug in connection tracking in the revSNAT flow: reply packets are
tracked before revSNAT restores the original tuple, meaning that it
won't match the existing CT entry, and a new one will be created.

This bug has impact on connections that set needs_ct (host-local and
egress gateway connections): packets coming in the reply direction won't
get the needs_ct mark and won't extend the expiration timeout.

The fix is to convert the tuple before doing the conntrack lookup, so
that it matches the tuple after revSNAT.

Fixes: #25110
Signed-off-by: Maxim Mikityanskiy <maxim@isovalent.com>

gentoo-root committed:
bpf: Remove repeating check that SNAT needs a CT entry

At the moment, both snat_v{4,6}_track_connection and
snat_v{4,6}_new_mapping check the same conditions to determine whether a
corresponding CT entry should exist for an SNAT entry. This approach has
the following drawbacks:

1. Duplicate code: risk of divergence and a minor performance penalty,
   because this code is actually executed twice for the first packet.

2. The checks in snat_v{4,6}_track_connection are not optimal, because
   this function is called from both egress and ingress flows, but the
   checks are only relevant for the egress flow (because the first
   packet is egress, and the subsequent packets will have the result of
   the checks cached in the SNAT state).

Solve both drawbacks by moving the checks one layer up, to
snat_v{4,6}_nat_handle_mapping (also put the checks to a separate
function for clarity). This way, the checks are only made in the egress
flow when the SNAT state doesn't exist yet, and the result is reused for
both snat_v{4,6}_track_connection and snat_v{4,6}_new_mapping, making
the code more robust and slightly more optimized.

Signed-off-by: Maxim Mikityanskiy <maxim@isovalent.com>
Suggested-by: Julian Wiedmann <jwi@isovalent.com>

gentoo-root committed:
bpf: Rename ct_lb_lookup4 to ct_lazy_lookup4 and add action parameter

Rename ct_lb_lookup4 to ct_lazy_lookup4, because it's going to be used
not only in LB flows, but now also in NAT. Add the action parameter to
cover ICMP packets in NAT flows.

The actual usage comes in the next commit, this one doesn't contain
functional changes.

Signed-off-by: Maxim Mikityanskiy <maxim@isovalent.com>

gentoo-root committed:
bpf: Reuse the NAT tuple for ct_lookup4

The tuple prepared by snat_v4_nat or snat_v4_rev_nat is good enough for
ct_lazy_lookup4, so we can spare a call to ct_extract_ports4, reducing
complexity, improving performance, but most importantly, allowing to
pass ports that differ from the packet header, which will allow to fix a
bug in a subsequent commit.

Signed-off-by: Maxim Mikityanskiy <maxim@isovalent.com>

asauber reviewed (approved):
CLI related changes LGTM

julianwiedmann commented on bpf/lib/nat.h:
```diff
@@ -1203,13 +1237,13 @@ snat_v4_rev_nat(struct __ctx_buff *ctx, const struct ipv4_nat_target *target, __
 		return ret;
 
-rewrite_ingress:
 	return snat_v4_rewrite_ingress(ctx, &tuple, state, off);
```
Note that this is currently missing a `has_l4_header` parameter. So I think it would attempt to revSNAT the L4 port even for a fragment.

gentoo-root replied:
Good point, thanks, it's a bug. As I'm splitting the pull requests, I think, it's more appropriate to fix it in the other pull request that fixes fragments.

(I'll also reuse has_l4_header for snat_v4_rewrite_egress instead of calculating it again.)

julianwiedmann commented on bpf/lib/nat.h:
```diff
@@ -338,9 +339,10 @@ static __always_inline int snat_v4_track_connection(struct __ctx_buff *ctx,
 	memcpy(&tmp, tuple, sizeof(tmp));
 
 	where = dir == NAT_DIR_INGRESS ? CT_INGRESS : CT_EGRESS;
+	ipv4_ct_tuple_swap_ports(&tmp);
```
Mind adding a short comment here why the swap is needed? (the callers extract the ports as-is, but CT wants tuples with swapped tuples)

gentoo-root replied:
This comment is already present inside the function body. Is it enough, or shall I add a small note also here, at the caller site?

julianwiedmann replied:
My thinking would be that it's good to have the calling path explicitly say "yes, this case applies to me". Just because all this stuff has proven to be super-tricky, and we're definitely going to get it wrong again in the future... :).

julianwiedmann commented on bpf/lib/nat.h:
```diff
@@ -338,9 +339,10 @@ static __always_inline int snat_v4_track_connection(struct __ctx_buff *ctx,
 
-	ret = ct_lookup4(get_ct_map4(&tmp), &tmp, ctx, off, where,
-			 &ct_state, &monitor);
+	ret = ct_lazy_lookup4(get_ct_map4(&tmp), &tmp, ctx, off, has_l4_header,
+			      ct_action, where, &ct_state, &monitor);
```
Mind splitting `bpf: Reuse the NAT tuple for ct_lookup4` into two patches - one trivial refactor that only does the rename & introduces the `action` parameter, and a second that actually adapts the IPv4 NAT code to use `ct_lazy_lookup4()` ?

I think that would make it a lot easier to follow what actually changed.

gentoo-root replied:
Done.

julianwiedmann commented on bpf/lib/nat.h:
```diff
@@ -403,7 +417,14 @@ snat_v4_rev_nat_handle_mapping(struct __ctx_buff *ctx,
+		tuple_revsnat.dport = (*state)->to_dport;
+	}
+
+	ret = snat_v4_track_connection(ctx, &tuple_revsnat, *state, has_l4_header, ct_action,
```
This chunk on its own (without the change to set `ostate->common.needs_ct = 1` for EgressGW connections) already fixes CT for replies of SNATed host connections, correct?

Then I would suggest to split this part off into its own patch, and describe the CT bug there in detail (as it actually pre-dates the EgressGW aspect). Right now the impact for such host connections isn't mentioned at all from what I can tell?

gentoo-root replied:
>This chunk on its own (without the change to set ostate->common.needs_ct = 1 for EgressGW connections) already fixes CT for replies of SNATed host connections, correct?

Yes.

>Then I would suggest to split this part off into its own patch, and describe the CT bug there in detail (as it actually pre-dates the EgressGW aspect).

Done.

julianwiedmann replied:
> > Then I would suggest to split this part off into its own patch, and describe the CT bug there in detail (as it actually pre-dates the EgressGW aspect).
> 
> Done.

:pray: Note that it would probably help readability if you flip the patch order around (ie. have `bpf: Use the right tuple to track connections on revSNAT` first). So that the EgressGW patch doesn't need to say
>The next commit is also required for the complete fix, otherwise the
needs_ct mark won't be matched.

, but simply sets `needs_ct`, updates the unit test and gets all the correct behaviour from the preceding fix.

julianwiedmann reviewed (changes requested):
Thank you! A few smaller change requests, mostly to highlight that we're fixing the CT lookup for more than "just" EgressGW replies. Would be good if you could adjust the PR title / description accordingly as well.

I don't quite follow why the first two patches in the series are needed - these are just cleanups, correct? Then I would suggest to split them into a follow-on PR, so this PR is easier to backport.

Note that we should also fix the IPv6 path (ie. have `snat_v6_rev_nat_handle_mapping()` revSNAT the tuple before the CT lookup), but that's perfectly fine as follow-on PR.

gentoo-root reviewed (commented):
> Thank you! A few smaller change requests, mostly to highlight that we're fixing the CT lookup for more than "just" EgressGW replies. Would be good if you could adjust the PR title / description accordingly as well.

Done.

> I don't quite follow why the first two patches in the series are needed - these are just cleanups, correct? Then I would suggest to split them into a follow-on PR, so this PR is easier to backport.

These were first steps in my bigger effort to unify the code. As it didn't happen, and there is indeed no dependency on the first two commits, I pushed them separately:

https://github.com/cilium/cilium/pull/25356

> Note that we should also fix the IPv6 path (ie. have `snat_v6_rev_nat_handle_mapping()` revSNAT the tuple before the CT lookup), but that's perfectly fine as follow-on PR.

Done here, no reason to postpone it.

julianwiedmann reviewed (changes requested):
Looks good! Just need to decide whether we add the full IPv6 fix here, or punt to a separate PR.

gentoo-root committed:
bpf: Rename ct_lb_lookup6 to ct_lazy_lookup6 and add action parameter

Rename ct_lb_lookup6 to ct_lazy_lookup6, because it's going to be used
not only in LB flows, but now also in NAT. Add the action parameter to
cover ICMP packets in NAT flows.

The actual usage comes in the next commit, this one doesn't contain
functional changes.

Signed-off-by: Maxim Mikityanskiy <maxim@isovalent.com>

gentoo-root committed:
bpf: Reuse the NAT tuple for ct_lookup6

The tuple prepared by snat_v6_nat or snat_v6_rev_nat is good enough for
ct_lazy_lookup6, so we can spare a call to ct_extract_ports6, reducing
complexity, improving performance, but most importantly, allowing to
pass ports that differ from the packet header, which will allow to fix a
bug in a subsequent commit.

Signed-off-by: Maxim Mikityanskiy <maxim@isovalent.com>

julianwiedmann reviewed (approved):
Looks all good, thanks a ton @gentoo-root !
